spark2/src/main/java/org/apache/iceberg/spark/SparkFilters.java

```diff
@@ -113,13 +115,13 @@ public static Expression convert(Filter filter) {
             // comparison with null in normal equality is always null. this is probably a mistake.
             Preconditions.checkNotNull(eq.value(),
                 "Expression is always false (eq is not null-safe): %s", filter);
-            return equal(eq.attribute(), convertLiteral(eq.value()));
+            return handleEqual(eq.attribute(), eq.value());
           } else {
             EqualNullSafe eq = (EqualNullSafe) filter;
             if (eq.value() == null) {
               return isNull(eq.attribute());
             } else {
-              return equal(eq.attribute(), convertLiteral(eq.value()));
+              return handleEqual(eq.attribute(), eq.value());
```

Comment:
why is this not directly inside `Expressions.equal`, so we can avoid duplication between spark 2 and 3?

Reply:
I thought to reject NaN in any predicate and let `SparkFilters` to do rewrites was the conclusion we reached in [this thread](https://github.com/apache/iceberg/pull/1747#discussion_r529096291)?

Reply:
Yes, I agree. Rewriting filters should be done in translation to Iceberg so that we have simpler behavior and strong assumptions.